File: frontend/src/services/voiceLive.js
// frontend/src/services/voiceLive.js
//
// Handles the full browser-side Voice Live pipeline:
//   1. Capture microphone audio
//   2. Convert it to 16-bit PCM at 24kHz (what Azure Voice Live expects)
//   3. Stream it over a WebSocket to the FastAPI backend (/ws/voice)
//   4. Receive audio replies + transcripts back, and play the audio
//
// Status values reported via onStatusChange, in order:
//   "connecting"  - WebSocket is being opened
//   "connected"   - socket open, mic is starting
//   "ready"       - backend confirmed the Voice Live session is configured;
//                    it is now safe to talk and audio will be sent
//   "error"       - something failed; onError carries a human-readable reason
//   "disconnected"- socket closed (either we called stop(), or it dropped)

const TARGET_SAMPLE_RATE = 24000;
const CONNECT_TIMEOUT_MS = 8000; // WebSocket must open within this long
const READY_TIMEOUT_MS = 12000; // backend must send {type:"ready"} within this long (mirrors the backend's own watchdog, with slack for network latency)

export class VoiceLiveClient {
  constructor({ wsUrl, onTranscript, onAssistantText, onError, onStatusChange, onToolActivity }) {
    this.wsUrl = wsUrl;
    this.onTranscript = onTranscript;
    this.onAssistantText = onAssistantText;
    this.onError = onError;
    this.onStatusChange = onStatusChange;
    // Fired when the backend calls one of the MCP tools (get_customer_ledger,
    // check_risk, ...) to ground its answer — lets the call UI show something
    // like "checking Utkarsh's ledger..." instead of a silent pause.
    // Payload: { type: "tool_call" | "tool_result", name, arguments?, data? }
    this.onToolActivity = onToolActivity;

    this.ws = null;
    this.sessionReady = false;
    this.stopped = false; // true once stop() has been called, so late/async
    // events from a closing socket or a dying AudioContext never reach the
    // UI as a spurious error after the person already ended the call.
    this.errored = false; // true once onError has fired, so the close that
    // inevitably follows (we always close the socket after an error)
    // doesn't overwrite the "error" status back to "disconnected" and wipe
    // out the error banner/retry button the person is looking at.

    this.connectTimer = null;
    this.readyTimer = null;

    // mic capture
    this.micContext = null;
    this.micSource = null;
    this.processor = null;
    this.muteGain = null;
    this.stream = null;

    // playback
    this.playbackContext = null;
    this.playbackQueueTime = 0;
  }

  _reportError(msg) {
    this.errored = true;
    this.onError?.(msg);
  }

  async start() {
    this.stopped = false;

    // Create (and try to resume) the playback AudioContext up front, in the
    // same call stack as the user's "Start a voice call" click, instead of
    // lazily on the first inbound audio chunk. A context created inside a
    // WebSocket message handler has no user-gesture "transient activation"
    // behind it, so browsers can leave it permanently suspended — audio
    // gets queued into it but never actually plays. That's the main reason
    // responses were sometimes silent. Creating + resuming it now, while
    // we're still inside the click's activation window, avoids that.
    try {
      this.playbackContext = new (window.AudioContext || window.webkitAudioContext)();
      this.playbackQueueTime = this.playbackContext.currentTime;
      if (this.playbackContext.state === "suspended") {
        await this.playbackContext.resume().catch(() => {});
      }
    } catch (err) {
      console.error("Could not create playback audio context:", err);
      // Not fatal by itself — we still try the call, but flag it so the
      // person understands why they might not hear anything.
      this.onError?.("Could not initialize audio playback in this browser.");
      this.errored = false; // recoverable — don't suppress a later real error
    }

    this.ws = new WebSocket(this.wsUrl);
    this.ws.binaryType = "arraybuffer";

    this.connectTimer = setTimeout(() => {
      if (this.ws?.readyState === WebSocket.CONNECTING) {
        this._reportError("Could not reach the server. Check your connection and that the backend is running.");
        this._hardStop("connect-timeout");
      }
    }, CONNECT_TIMEOUT_MS);

    this.ws.onopen = () => {
      clearTimeout(this.connectTimer);
      this.onStatusChange?.("connected");

      // Once the session is configured Azure-side, the backend must tell us
      // — if it never does (Azure unreachable, bad credentials that only
      // fail after connecting, etc.) don't leave the call UI stuck forever.
      this.readyTimer = setTimeout(() => {
        if (!this.sessionReady && !this.stopped) {
          this._reportError("Voice Live didn't respond in time. Please try again.");
          this._hardStop("ready-timeout");
        }
      }, READY_TIMEOUT_MS);

      this._startMic().catch((err) => {
        console.error("Mic error:", err);
        this._reportError("Could not access microphone: " + err.message);
        // Without a mic there's nothing useful this call can do — end it
        // cleanly instead of leaving a connected-but-mute socket open,
        // which is exactly what made "Start Voice Call" look like it did
        // nothing.
        this._hardStop("mic-failed");
      });
    };

    this.ws.onmessage = (event) => {
      if (this.stopped) return;
      if (typeof event.data === "string") {
        this._handleJson(event.data);
      } else {
        this._playAudioChunk(event.data);
      }
    };

    this.ws.onerror = () => {
      if (this.stopped) return;
      // The browser gives us no detail on WebSocket error events; the
      // close event that immediately follows carries the actual reason
      // (see onclose below), so we only use this as a fallback signal.
      this._reportError("WebSocket connection error");
    };

    this.ws.onclose = (event) => {
      clearTimeout(this.connectTimer);
      clearTimeout(this.readyTimer);
      const wasStopped = this.stopped;
      const alreadyErrored = this.errored;
      this._stopMic();
      this._stopPlayback();

      // A clean, user-initiated close (code 1000, or stop() already ran)
      // needs no explanation. Anything else — the backend rejecting us,
      // Azure dropping, a network blip — should be visible instead of
      // just silently going quiet.
      if (!wasStopped && !alreadyErrored && event.code !== 1000) {
        this._reportError(
          event.reason || `Connection closed unexpectedly (code ${event.code}).`
        );
      }

      // Don't downgrade an "error" status the UI is already showing (with
      // its retry button) back to a plain "disconnected" — the close that
      // follows almost every error is expected, not new information.
      if (!this.errored) {
        this.onStatusChange?.("disconnected");
      }
    };
  }

  stop() {
    this._hardStop("user-ended");
  }

  _hardStop(_reason) {
    this.stopped = true;
    clearTimeout(this.connectTimer);
    clearTimeout(this.readyTimer);
    try {
      this.ws?.close(1000, "client stopped");
    } catch {
      // ignore — socket may already be closed/closing
    }
    this._stopMic();
    this._stopPlayback();
  }

  _handleJson(raw) {
    let payload;
    try {
      payload = JSON.parse(raw);
    } catch {
      return;
    }

    if (payload.type === "ready") {
      clearTimeout(this.readyTimer);
      this.sessionReady = true;
      this.onStatusChange?.("ready");
    } else if (payload.type === "transcript") {
      this.onTranscript?.(payload.text);
    } else if (payload.type === "assistant_text") {
      this.onAssistantText?.(payload.text);
    } else if (payload.type === "tool_call" || payload.type === "tool_result") {
      this.onToolActivity?.(payload);
    } else if (payload.type === "error") {
      this._reportError(payload.message);
    }
  }

  async _startMic() {
    this.stream = await navigator.mediaDevices.getUserMedia({ audio: true });
    if (this.stopped) {
      // stop() was called while the permission prompt was open — tear the
      // stream back down instead of wiring it up.
      this.stream.getTracks().forEach((t) => t.stop());
      this.stream = null;
      return;
    }

    this.micContext = new (window.AudioContext || window.webkitAudioContext)();
    if (this.micContext.state === "suspended") {
      await this.micContext.resume().catch(() => {});
    }
    this.micSource = this.micContext.createMediaStreamSource(this.stream);

    // ScriptProcessorNode is deprecated but still universally supported and
    // far simpler to wire up than AudioWorklet for a student project deadline.
    const bufferSize = 4096;
    this.processor = this.micContext.createScriptProcessor(bufferSize, 1, 1);

    const inputSampleRate = this.micContext.sampleRate; // usually 48000

    this.processor.onaudioprocess = (e) => {
      if (this.ws?.readyState !== WebSocket.OPEN) return;
      // Don't stream audio to Azure until the backend has confirmed the
      // session is configured (voice/instructions/VAD/tools). Sending
      // earlier meant Azure could process the first moments of speech
      // under stale/default config — a real cause of calls that seemed to
      // "do nothing" or reply in the wrong language.
      if (!this.sessionReady) return;

      const input = e.inputBuffer.getChannelData(0);
      const resampled = downsample(input, inputSampleRate, TARGET_SAMPLE_RATE);
      const pcm16 = floatTo16BitPCM(resampled);
      this.ws.send(pcm16.buffer);
    };

    this.micSource.connect(this.processor);

    // IMPORTANT: ScriptProcessorNode only fires onaudioprocess if connected
    // to a destination — but we do NOT want to hear our own raw mic audio
    // played back (that would cause an instant feedback echo). Route it
    // through a silent gain node instead of straight to speakers.
    this.muteGain = this.micContext.createGain();
    this.muteGain.gain.value = 0;
    this.processor.connect(this.muteGain);
    this.muteGain.connect(this.micContext.destination);
  }

  _stopMic() {
    this.processor?.disconnect();
    this.micSource?.disconnect();
    this.muteGain?.disconnect();
    this.stream?.getTracks().forEach((t) => t.stop());
    this.micContext?.close().catch(() => {});

    this.processor = null;
    this.micSource = null;
    this.muteGain = null;
    this.stream = null;
    this.micContext = null;
  }

  _stopPlayback() {
    // Previously the playback AudioContext was never closed, so every call
    // left one behind. Browsers cap the number of live AudioContexts, so a
    // few calls in the same session could start silently failing to play
    // anything at all.
    this.playbackContext?.close().catch(() => {});
    this.playbackContext = null;
    this.playbackQueueTime = 0;
  }

  _playAudioChunk(arrayBuffer) {
    if (!this.playbackContext) return; // torn down (call ended) or failed to init

    if (this.playbackContext.state === "suspended") {
      // Best-effort re-resume — some browsers suspend a context again after
      // periods of silence.
      this.playbackContext.resume().catch(() => {});
    }

    const int16 = new Int16Array(arrayBuffer);
    const float32 = new Float32Array(int16.length);
    for (let i = 0; i < int16.length; i++) {
      float32[i] = int16[i] / 32768;
    }

    const audioBuffer = this.playbackContext.createBuffer(1, float32.length, TARGET_SAMPLE_RATE);
    audioBuffer.copyToChannel(float32, 0);

    const src = this.playbackContext.createBufferSource();
    src.buffer = audioBuffer;
    src.connect(this.playbackContext.destination);

    // queue chunks back-to-back instead of overlapping them
    const startAt = Math.max(this.playbackQueueTime, this.playbackContext.currentTime);
    src.start(startAt);
    this.playbackQueueTime = startAt + audioBuffer.duration;
  }
}

function downsample(buffer, inputRate, outputRate) {
  if (outputRate === inputRate) return buffer;
  const ratio = inputRate / outputRate;
  const newLength = Math.round(buffer.length / ratio);
  const result = new Float32Array(newLength);
  for (let i = 0; i < newLength; i++) {
    result[i] = buffer[Math.floor(i * ratio)];
  }
  return result;
}

function floatTo16BitPCM(float32Array) {
  const out = new Int16Array(float32Array.length);
  for (let i = 0; i < float32Array.length; i++) {
    const s = Math.max(-1, Math.min(1, float32Array[i]));
    out[i] = s < 0 ? s * 0x8000 : s * 0x7fff;
  }
  return out;
}

File: backend/app/routers/voice_live.py
# WebSocket bridge: Browser mic  <->  FastAPI  <->  Azure Voice Live
#
# Replaces pyaudio (physical mic/speaker) with a WebSocket connection
# to the browser. This is the ASK pipeline — full-duplex conversational
# voice, grounded in real ledger data via the MCP tools below. (Entries
# use a separate, lighter batch-STT pipeline — see routers/transcribe.py
# and routers/transactions.py — Voice Live is deliberately not used there.)
#
# Frontend connects to:  ws://localhost:8000/ws/voice
# Frontend sends:   binary WebSocket messages, each one a chunk of
#                    16-bit PCM audio at 24kHz, mono (same format
#                    Voice Live expects from pyaudio today)
# Frontend receives: binary WebSocket messages = audio to play back,
#                     plus JSON messages for transcript/tool/error events.

import asyncio
import base64
import contextlib
import json
import logging
import os

import sys
from pathlib import Path

from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from dotenv import load_dotenv

from azure.core.credentials import AzureKeyCredential
from azure.ai.voicelive.aio import connect
from azure.ai.voicelive.models import (
    AudioEchoCancellation,
    AudioNoiseReduction,
    AzureStandardVoice,
    AudioInputTranscriptionOptions,
    FunctionCallOutputItem,
    FunctionTool,
    InputAudioFormat,
    Modality,
    OutputAudioFormat,
    RequestSession,
    ServerEventType,
    AzureSemanticVadMultilingual,
)

# Ensure project root is in sys.path when running from backend directory
# (same pattern as routers/agent.py, since this also calls into agent/).
PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from agent.mcp_server import list_tools, call_tool
from agent.agent_config import get_system_prompt

load_dotenv()
router = APIRouter()
logger = logging.getLogger("voiceledger.voice_live")

# How long we'll wait for Azure to confirm our session.update() before
# giving up and telling the browser something is wrong, instead of leaving
# the call UI stuck forever.
SESSION_READY_TIMEOUT_S = 10


def build_tools() -> list:
    """Expose the same MCP tools the text agent uses (agent/mcp_server.py),
    so Voice Live answers are grounded in real Supabase data instead of the
    model improvising a number."""
    return [
        FunctionTool(
            name=tool["name"],
            description=tool["description"],
            parameters=tool["inputSchema"],
        )
        for tool in list_tools()
    ]


def build_session() -> RequestSession:
    """Same session config as voice_assistant.py, plus the MCP tools."""
    instructions = (
        get_system_prompt()
        + "\n\n## Voice call rules\n"
        + "You are speaking with the shopkeeper live, over voice, in a phone call. "
        "You understand Hindi, English and Hinglish. "
        "LANGUAGE RULE: If the user speaks Hindi, answer in Hindi. "
        "If the user speaks English, answer in English. "
        "If the user speaks Hinglish, answer naturally in Hinglish. "
        "Never translate Hindi to English unless explicitly asked. "
        "Keep answers short and conversational, like a real spoken reply — "
        "no bullet points, no reading out raw JSON. "
        "Always call the relevant tool before answering a question about a "
        "customer, balance, risk, or reminder — never guess a number."
    )

    return RequestSession(
        modalities=[Modality.TEXT, Modality.AUDIO],
        instructions=instructions,
        # "en-US-Ava:DragonHDLatestNeural" is a single-language HD voice —
        # it does not officially support code-switching, which is the root
        # cause of the "language switches unexpectedly" bug: when the model
        # emits Hindi/Hinglish text, that voice's language auto-detection is
        # unreliable and it drifts back to (or randomly switches into)
        # English. "en-US-AvaMultilingualNeural" is the multilingual
        # counterpart of the same voice character, built to speak whichever
        # language the text is in. We leave `locale` unset so it keeps
        # auto-detecting per sentence (setting it would force one language
        # and mute the others), and bias the English accent toward Indian
        # English via prefer_locales without touching Hindi output.
        voice=AzureStandardVoice(
            name="en-US-AvaMultilingualNeural",
            prefer_locales=["en-IN"],
        ),
        input_audio_format=InputAudioFormat.PCM16,
        output_audio_format=OutputAudioFormat.PCM16,
        input_audio_transcription=AudioInputTranscriptionOptions(
            model="azure-speech",
            language="hi-IN,en-IN",
            phrase_list=["VoiceLedger", "udhaar", "baaki", "hisaab"],
        ),
        turn_detection=AzureSemanticVadMultilingual(
            threshold=0.5,
            prefix_padding_ms=500,
            silence_duration_ms=700,
            speech_duration_ms=100,
            create_response=True,
            interrupt_response=True,
        ),
        input_audio_echo_cancellation=AudioEchoCancellation(),
        input_audio_noise_reduction=AudioNoiseReduction(type="azure_deep_noise_suppression"),
        tools=build_tools(),
        tool_choice="auto",
    )


async def _handle_tool_call(connection, websocket: WebSocket, event, shopkeeper_id: str) -> None:
    """A function_call_arguments.done event means the model wants to call
    one of our MCP tools. Run it for real, tell the browser (so the call UI
    can show "checking Utkarsh's ledger..."), then feed the result back into
    the session so the model's spoken answer is grounded in it."""
    try:
        arguments = json.loads(event.arguments) if event.arguments else {}
    except json.JSONDecodeError:
        arguments = {}

    await websocket.send_json({
        "type": "tool_call",
        "name": event.name,
        "arguments": arguments,
    })

    # The MCP tools use the synchronous Supabase client — run off the event
    # loop thread so a slow DB call doesn't stall the audio stream. If the
    # tool itself raises (bad args, DB hiccup, etc.) we still have to answer
    # Azure's function call with *something*, or the model just hangs
    # waiting for output and the call goes silent — so we feed back an
    # error payload instead of letting the exception escape.
    try:
        result = await asyncio.to_thread(call_tool, event.name, arguments, shopkeeper_id)
    except Exception as exc:  # noqa: BLE001 - deliberately broad, see above
        logger.exception("Tool call %s failed", event.name)
        result = {"error": f"{event.name} failed: {exc}"}

    await websocket.send_json({
        "type": "tool_result",
        "name": event.name,
        "data": result,
    })

    await connection.conversation.item.create(
        item=FunctionCallOutputItem(
            call_id=event.call_id,
            output=json.dumps(result, default=str),
        )
    )
    await connection.response.create()


async def _safe_send_json(websocket: WebSocket, payload: dict) -> None:
    """Best-effort JSON send — the browser side may already be gone by the
    time we try to report an error, and that must never itself raise."""
    with contextlib.suppress(Exception):
        await websocket.send_json(payload)


async def _fail(websocket: WebSocket, message: str, code: int = 1011) -> None:
    """Tell the browser clearly why the call can't proceed, THEN close.
    Closing first (the old behavior) throws away the reason: the frontend's
    onclose handler only sees a numeric close code, not `message`, so every
    setup failure looked like a silent hang."""
    logger.warning("Voice Live call failed: %s", message)
    await _safe_send_json(websocket, {"type": "error", "message": message})
    with contextlib.suppress(Exception):
        await websocket.close(code=code, reason=message[:120])


@router.websocket("/ws/voice")
async def voice_ws(websocket: WebSocket):
    await websocket.accept()

    # Root cause of the ledger/customer lookup failures: shopkeeper_id was
    # never read from the connection, so every shopkeeper-scoped tool call
    # (get_customer_ledger, check_risk, generate_reminder,
    # get_business_summary) fell back to call_tool()'s shopkeeper_id=None
    # default and returned {"error": "No shopkeeper session — cannot look
    # up data."} — which the model then turned into a graceful-sounding
    # apology. The frontend sends it as a query param (see
    # frontend/src/components/AskAssistant.jsx); everything else about the
    # audio/WebSocket pipeline is unchanged.
    shopkeeper_id = websocket.query_params.get("shopkeeper_id")
    if not shopkeeper_id:
        await _fail(websocket, "Missing shopkeeper_id — cannot start a scoped Voice Live session.")
        return

    endpoint = os.getenv("AZURE_VOICELIVE_ENDPOINT")
    api_key = os.getenv("AZURE_VOICELIVE_API_KEY")
    model = os.getenv("AZURE_VOICELIVE_MODEL", "gpt-4o")
    api_version = os.getenv("AZURE_VOICELIVE_API_VERSION", "2026-04-10")

    if not endpoint or not api_key:
        await _fail(websocket, "Azure Voice Live credentials are missing on the server "
                                "(AZURE_VOICELIVE_ENDPOINT / AZURE_VOICELIVE_API_KEY).")
        return

    try:
        async with connect(
            endpoint=endpoint,
            credential=AzureKeyCredential(api_key),
            model=model,
            api_version=api_version,
        ) as connection:
            await _run_call(connection, websocket, shopkeeper_id)
    except WebSocketDisconnect:
        pass
    except Exception as exc:  # noqa: BLE001 - connect()/session setup failed
        # This is what used to fail silently: a bad endpoint, an expired
        # key, an unsupported model/api-version, or Azure being unreachable
        # all raised here, FastAPI closed the socket with no explanation,
        # and the browser just saw the call go dead.
        logger.exception("Voice Live connection setup failed")
        await _fail(websocket, f"Could not start the Voice Live session: {exc}")


async def _run_call(connection, websocket: WebSocket, shopkeeper_id: str) -> None:
    """Runs one voice call end-to-end: configures the session, then bridges
    browser <-> Azure audio/events until either side disconnects."""

    session_ready = asyncio.Event()

    async def receive_from_browser():
        """Browser mic chunks -> Azure"""
        try:
            while True:
                chunk = await websocket.receive_bytes()
                # Wait for Azure to confirm session.update() before forwarding
                # any audio. session.update() only *sends* the config — it
                # does not wait for the session.updated ack — so without this
                # gate, mic audio captured right after the socket opens can
                # reach Azure before our instructions/voice/VAD/tools are
                # applied. Azure then processes it under stale/default
                # config, which is what caused "sometimes nothing happens",
                # "no response audio", and inconsistent language switching.
                await session_ready.wait()
                audio_b64 = base64.b64encode(chunk).decode("utf-8")
                await connection.input_audio_buffer.append(audio=audio_b64)
        except WebSocketDisconnect:
            pass
        except RuntimeError:
            # receive_bytes() called after the socket already closed.
            pass

    async def send_to_browser():
        """Azure audio replies + tool grounding -> browser"""
        try:
            async for event in connection:
                # print("AZURE EVENT:", event.type)
                if event.type == ServerEventType.SESSION_UPDATED:
                    if not session_ready.is_set():
                        session_ready.set()
                        # Tells the frontend it's now safe to start streaming
                        # mic audio and that the UI can leave "connecting".
                        await websocket.send_json({"type": "ready"})

                elif event.type == ServerEventType.RESPONSE_AUDIO_DELTA:
                    await websocket.send_bytes(event.delta)

                elif event.type == ServerEventType.CONVERSATION_ITEM_INPUT_AUDIO_TRANSCRIPTION_COMPLETED:
                    transcript = getattr(event, "transcript", "")
                    await websocket.send_json({"type": "transcript", "text": transcript})

                elif event.type == ServerEventType.RESPONSE_AUDIO_TRANSCRIPT_DONE:
                    text = getattr(event, "transcript", "")
                    await websocket.send_json({"type": "assistant_text", "text": text})

                elif event.type == ServerEventType.RESPONSE_FUNCTION_CALL_ARGUMENTS_DONE:
                    await _handle_tool_call(connection, websocket, event, shopkeeper_id)

                elif event.type == ServerEventType.ERROR:
                    msg = getattr(event.error, "message", str(event.error))
                    await websocket.send_json({"type": "error", "message": msg})
        except WebSocketDisconnect:
            pass
        except Exception as exc:  # noqa: BLE001 - Azure connection dropped/errored mid-call
            logger.exception("Voice Live event loop failed")
            await _safe_send_json(websocket, {"type": "error", "message": f"Voice Live connection lost: {exc}"})

    async def watch_session_ready():
        """If Azure never acks session.update(), the UI would otherwise sit
        on "connecting" forever with no explanation. Time it out instead by
        reporting the problem and closing the socket, which naturally ends
        receive_from_browser/send_to_browser too. NOTE: this task must NOT
        be raced with FIRST_COMPLETED alongside the other two — it finishes
        successfully (silently) the moment the session becomes ready, which
        would otherwise be mistaken for "the call ended" and tear everything
        down right as it starts."""
        try:
            await asyncio.wait_for(session_ready.wait(), timeout=SESSION_READY_TIMEOUT_S)
        except asyncio.TimeoutError:
            await _safe_send_json(websocket, {
                "type": "error",
                "message": "Voice Live didn't confirm the session in time. Please try again.",
            })
            with contextlib.suppress(Exception):
                await websocket.close(code=1011, reason="session not ready in time")

    await connection.session.update(session=build_session())

    # run mic streaming and Azure event handling concurrently; stop both as
    # soon as either one ends. The readiness watchdog runs alongside them
    # but independently (see its docstring above) and is cancelled in the
    # `finally` below rather than raced here.
    receiver = asyncio.create_task(receive_from_browser())
    sender = asyncio.create_task(send_to_browser())
    watchdog = asyncio.create_task(watch_session_ready())

    try:
        done, pending = await asyncio.wait({receiver, sender}, return_when=asyncio.FIRST_COMPLETED)
        for task in pending:
            task.cancel()
        for task in pending:
            with contextlib.suppress(asyncio.CancelledError):
                await task
        # surface exceptions from whichever task finished first, instead of
        # swallowing them (the old code never checked task results at all)
        for task in done:
            exc = task.exception() if not task.cancelled() else None
            if exc and not isinstance(exc, WebSocketDisconnect):
                raise exc
    finally:
        watchdog.cancel()
        with contextlib.suppress(asyncio.CancelledError, Exception):
            await watchdog
        with contextlib.suppress(Exception):
            await websocket.close()